### ISSUE TITLE
Support spec with finalizers in namespace resource

### ISSUE THREAD

**Issue by frvge** [hashicorp/terraform-provider-kubernetes#855]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In #701 , namespace data sources were added that can handle finalizers. The resource doesn't have support for it yet as far as I can tell from the code and the documentation.

### Potential Terraform Configuration

```hcl
resource "kubernetes_namespace" "test" {
  metadata {
    name = "mytest"
  }
  spec {
    finalizers = ['myfinalizer0','myfinalizer1']
  }
}
```

### References

- #701 
- https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers
- https://github.com/kubernetes/kubernetes/issues/77086

**Comment by dak1n1** [comment]
I support the idea of adding finalizers to the provider, since they can be useful for keeping resources around until deletion is needed. Ideally we would implement this generically, at the metadata level, so that we don't have to add it to each applicable spec. That way it can be applied to many types of resources.

Here are the API docs showing finalizers at the metdata level: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#objectmeta-v1-meta

And while you can find some resources with finalizers in the spec, I think the metadata approach would give us the most benefit and coverage of resource types.

Here's one test demonstrating using a random finalizer on a random new namespace, to show that metadata finalizers do an adequate job for a common, simple use case:

```
[dakini@dax 20200819]$ kube create namespace potato
namespace/potato created
[dakini@dax 20200819]$ kube edit namespace potato
namespace/potato edited
[dakini@dax 20200819]$ kube get namespace potato -o yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: "2020-08-19T21:24:32Z"
  finalizers:
  - kubernetes.io/potato   ####### random finalizer added  here
  managedFields:
...
  name: potato
...
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
[dakini@dax 20200819]$ kube delete namespace potato
namespace "potato" deleted

# deletion hangs. Here's a view of the namespace from another window
[dakini@dax ~]$ kube get namespace potato
NAME     STATUS        AGE
potato   Terminating   111s

# the finalizer is doing its job by preserving the namespace
```

There is some information here about the difference between namespace spec finalizers and the more generic metadata finalizers, for reference. https://github.com/kubernetes/kubernetes/issues/65994#issuecomment-404875556

So to summarize, I support adding finalizers specifically at the metadata level. And since there is a PR open for the spec level finalizer, we can go ahead with that. I just wouldn't recommend following this pattern for all the other resource types, unless there is a specific use case where spec level finalizers are needed.

**Comment by dak1n1** [comment]
We explored the solution here https://github.com/hashicorp/terraform-provider-kubernetes/pull/950#issuecomment-710215670 and found that implementing the namespace.spec finalizer is actually not feasible in the provider. (It's possible to do, but it will have side effects that users will probably not want, such as deleting all resources inside of the namespace when adding/removing finalizers). However, I've opened a new issue here https://github.com/hashicorp/terraform-provider-kubernetes/issues/1040 for the metadata level finalizer. This one will actually allow for updates and it should be a lot more useful. And it will be usable on all resources, instead of just the namespace resource. Please vote on this issue if you want to see it implemented. Thanks! I'll go ahead and close this one in favor of the new issue.
